butterknife-compiler/src/main/java/butterknife/compiler/ButterKnifeProcessor.java

```diff
@@ -143,10 +149,18 @@
     debuggable = !"false".equals(env.getOptions().get(OPTION_DEBUGGABLE));
 
     typeUtils = env.getTypeUtils();
+    elementUtils = env.getElementUtils();
     filer = env.getFiler();
     try {
-      trees = Trees.instance(processingEnv);
-    } catch (IllegalArgumentException ignored) {
+      //reflection won't be necessary after https://github.com/gradle/gradle/pull/8393
+      java.lang.reflect.Field delegateField = processingEnv.getClass().getDeclaredField("delegate");
+      delegateField.setAccessible(true);
+      trees = Trees.instance((ProcessingEnvironment) delegateField.get(processingEnv));
+    } catch (Throwable t) {
+      try {
+        trees = Trees.instance(processingEnv);
+      } catch (Throwable ignored) {
```

Comment:
This whole block here is meant to access the `ProcessingEnvironment` instance passed by gradle. Gradle wraps the annotation processors internally so that it can control the filer passed to annotation processors. 

The wrapper class provided by gradle is `IncrementalProcessingEnvironment`. This class is not exposed by Gradle. The idea is to access this class and the original wrapped `ProcessingEnvironment`. We need this code because `Trees.getInstance(..)` checks that it receives the original processing environment. If you look at the code of the `Trees` class, you will see that it first checks the class name of the instance. I tried to by pass this check and call the method that `instance` is calling internally. But there are additional checks later on the road. I resorted to use reflection to access the original processing environment in the non exposed gradle wrapper class.

I have opened an issue in gradle's code base on this in: https://github.com/gradle/gradle/pull/8393
The PR has been rejected has gradle doesn't want to expose this class. As you can see in this issue, we discuss the safety of accessig `Trees` in the BK processor during incremental annotation processing. This practice is a corner case when incremental processing is used and it is generally not advised to turn annotation processors to incremental if they use this API. Though, I explained why I believe there is no issue in doing so in our case.

Also, at the end of the discussion in the gradle PR, gradle says that the processing environment is always wrapped. It was not my experience when testing: I received the vanilla processing environment. I didn't follow up with gradle on this. But for this PR, I don't think it should matter as the case was already handled by just swalling the exception. In practice, the PR has been tested at an industrial scale in 2 companies that I know of and all works well.